Enable Strictly Safe Swift in WebKit

adetaylor opened the pull request:
#### 6bce8081f2644bee7f9510cc662b312b7c8a8bd3
<pre>
Enable Strictly Safe Swift in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=296428">https://bugs.webkit.org/show_bug.cgi?id=296428</a>
<a href="https://rdar.apple.com/156597508">rdar://156597508</a>

Reviewed by David Kilzer and Elliott Williams.

Enables Strictly Safe Swift errors across the WebKit project.

* Source/WebKit/Configurations/WebKit.xcconfig:

Canonical link: <a href="https://commits.webkit.org/298838@main">https://commits.webkit.org/298838@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f1b08297deda445087820fbfb998963b7454365

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122889 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67401 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a3b7ef22-14f8-4372-a09a-29ead712c741) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45078 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88705 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43114 "Found 2 new test failures: fast/css/vertical-align-rem-on-root.html storage/indexeddb/aborted-versionchange-closes.html (failure)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8b3091ae-482e-4301-9707-b033c91950d0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119762 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29641 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104803 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69174 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28709 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66554 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99020 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23063 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126023 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32830 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97371 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44076 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101004 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97169 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24745 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42496 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20455 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39720 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43598 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49194 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43065 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46404 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44770 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -65,6 +65,24 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
+SWIFT_STRICT_MEMORY_SAFETY_MACOS_SINCE_2600 = YES;
+SWIFT_STRICT_MEMORY_SAFETY_IOS_BEFORE_26 = NO;
+SWIFT_STRICT_MEMORY_SAFETY_IOS_SINCE_26 = YES;
+OTHER_SWIFT_FLAGS = $(inherited);
```
won't this conflict / override / be overridden by the other OTHER_SWIFT_FLAGS line I added in this file?

adetaylor replied:
Ah yeah, I'll fix that

rr-codes commented on Source/WebKit/UIProcess/Cocoa/WKURLSchemeHandlerAdapter.swift:
```diff
@@ -38,7 +38,12 @@ final class WKURLSchemeHandlerAdapter: NSObject, WKURLSchemeHandler {
         let task = Task {
             do {
-                for try await result in wrapped.reply(for: urlSchemeTask.request) {
+                // Safety:
```
where is the unsafe keyword here?

rr-codes replied:
oh it's there I am just blind

emw-apple commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -65,6 +65,21 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
+WK_SWIFT_WERROR_FLAGS[sdk=macosx*] = $(WK_SWIFT_WERROR_FLAGS$(WK_MACOS_2600));
+WK_SWIFT_WERROR_FLAGS[sdk=iphone*] = $(WK_SWIFT_WERROR_FLAGS$(WK_IOS_26));
+WK_SWIFT_WERROR_FLAGS_MACOS_SINCE_2600 = -Werror StrictMemorySafety;
+WK_SWIFT_WERROR_FLAGS_SINCE_26 = -Werror StrictMemorySafety;
```
```suggestion
WK_SWIFT_WERROR_FLAGS_IOS_SINCE_26 = -Werror StrictMemorySafety;
```

emw-apple commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -65,6 +65,21 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
+SWIFT_STRICT_MEMORY_SAFETY_MACOS_BEFORE_2600 = NO;
+SWIFT_STRICT_MEMORY_SAFETY_MACOS_SINCE_2600 = YES;
+SWIFT_STRICT_MEMORY_SAFETY_IOS_BEFORE_26 = NO;
+SWIFT_STRICT_MEMORY_SAFETY_IOS_SINCE_26 = YES;
```
You can remove all the "NO" bindings here; usually we use the empty string to fall back to the default behavior. This is also kind of nice because it turns down the number of "NO" environment variables we set in every build script and spew into logs.

```suggestion
SWIFT_STRICT_MEMORY_SAFETY[sdk=macosx*] = $(SWIFT_STRICT_MEMORY_SAFETY$(WK_MACOS_2600));
SWIFT_STRICT_MEMORY_SAFETY[sdk=iphone*] = $(SWIFT_STRICT_MEMORY_SAFETY$(WK_IOS_26));
SWIFT_STRICT_MEMORY_SAFETY_MACOS_SINCE_2600 = YES;
SWIFT_STRICT_MEMORY_SAFETY_IOS_SINCE_26 = YES;
```

ddkilzer commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -65,6 +65,18 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
+SWIFT_STRICT_MEMORY_SAFETY[sdk=macosx*] = $(SWIFT_STRICT_MEMORY_SAFETY$(WK_MACOS_2600));
+SWIFT_STRICT_MEMORY_SAFETY[sdk=iphone*] = $(SWIFT_STRICT_MEMORY_SAFETY$(WK_IOS_26));
+SWIFT_STRICT_MEMORY_SAFETY_MACOS_SINCE_2600 = YES;
+SWIFT_STRICT_MEMORY_SAFETY_IOS_SINCE_26 = YES;
```
Why is this based on the OS version and not the Xcode version?

Why aren't we applying this to tvOS, visionOS and watchOS as well?

emw-apple replied:
+1. This will also be a problem when we switch to a back-deployed macOS downlevel build configuration. We need to derive this setting from SDK versions or Xcode version, since those are the version numbers associated with the compiler toolchain.

adetaylor replied:
I've switched to the Xcode version. I can't exactly remember why I did it based on OS version - I suspect it was just for consistency with other places we conditionalise Swift stuff but I agree the Xcode/SDK version is better.

ddkilzer commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -65,6 +65,18 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
+WK_SWIFT_WERROR_FLAGS[sdk=macosx*] = $(WK_SWIFT_WERROR_FLAGS$(WK_MACOS_2600));
+WK_SWIFT_WERROR_FLAGS[sdk=iphone*] = $(WK_SWIFT_WERROR_FLAGS$(WK_IOS_26));
+WK_SWIFT_WERROR_FLAGS_MACOS_SINCE_2600 = -Werror StrictMemorySafety;
+WK_SWIFT_WERROR_FLAGS_IOS_SINCE_26 = -Werror StrictMemorySafety;
```
Same questions as above:

Why is this based on the OS version and not the Xcode version?

Why aren't we applying this to tvOS, visionOS and watchOS as well?

adetaylor replied:
Adjusted thusly.

ddkilzer reviewed (approved):
This seems fine to land as-is (to prevent future regressions), but I'd like to understand:
- Why we're using OS version instead of Xcode version.
- Why we're not applying this to tvOS, visionOS and watchOS.

emw-apple commented on Source/WebKit/UIProcess/Cocoa/WKURLSchemeHandlerAdapter.swift:
```diff
@@ -38,7 +38,12 @@ final class WKURLSchemeHandlerAdapter: NSObject, WKURLSchemeHandler {
+                // This is believed to be a false-positive and this code is actually safe;
+                // a future Swift compiler will remove the need for the 'unsafe' keyword
+                // here.
+                for try await unsafe result in wrapped.reply(for: urlSchemeTask.request) {
```
Out of curiosity: With `-Werror StrictMemorySafety`, once we get a build of the compiler with this false-positive fixed, will this use site become an error?

adetaylor replied:
This seems to have been fixed in recent builds so I am going to remove this change. But that also allows me to answer your question: no error was emitted about unnecessary unsafe use.

rr-codes commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -65,6 +65,12 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
 
+// Enable strict memory safety in Swift, and treat any such warnings as errors.
+// In time this should be promoted to CommonBase.xcconfig so it applies more broadly.
+// Enable for sufficiently recent XCode only.
```
this is the most minor nit ever, but it's "Xcode" and not "XCode" 😛

adetaylor replied:
FIxed. THanks :)

adetaylor committed:
Enable Strictly Safe Swift in WebKit
https://bugs.webkit.org/show_bug.cgi?id=296428
rdar://156597508

Reviewed by David Kilzer and Elliott Williams.

Enables Strictly Safe Swift errors across the WebKit project.

* Source/WebKit/Configurations/WebKit.xcconfig:

Canonical link: https://commits.webkit.org/298838@main
